source/common/stats/stats_impl.cc

```diff
@@ -150,10 +150,21 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
 }
 
 RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
-  // This must be zero-initialized
   RawStatData* data = static_cast<RawStatData*>(::calloc(RawStatData::size(), 1));
   data->initialize(name);
-  return data;
+
+  std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
Good question.  Fundamentally it doesn't have to be this way, but this is an artifact of the way STL sets & maps work.  STL set lookups require construction of an object.  If you make this an `map<string, RawStatData*>` you have just pushed the problem around a little, as you'd need to copy the string to potentially truncate it, which is the same work, basically, as is being done here, and then you have to duplicate the truncation logic instead of just having it in RawStatData::initialize.  Worse, you'd wind up permanently duplicating all the name storage.  I argued that I don't know really how impactful that would be across different ways you might scale the system, but the current solution has zero overhead from duplication and is really no more complex from a programming perspective.

One question to ask is whether RawStatData::initialize is doing anything extra that's not required for the set lookup.   It is, but it's pretty minimal and IMO not worth optimizing around.

An ideal solution would allow the set lookup against a string_view, without actually constructing the templated type.  BlockMemoryHashSet::insert has that signature, so in the hot-restart case you don't need to do the prospective allocation.

Reply:
Yeah, fair enough. Maybe add a comment to the code capturing this design history. Thanks!